### PR TITLE
Utkarsha/dera 453/websocket connection closed due to timeout

### PR DESCRIPTION
**Pull request by utkarsha-deriv**
This PR helps to reconnect the websocket back to the server incase the server goes down.

### CODE REVIEW

File: src/configs/websocket/index.ts [hunk 1]
```diff
@@ -15,12 +15,17 @@ export type TDerivApi = {
...
+const RECONNECT_INTERVAL = attempts * 10000;
 const PING_INTERVAL = 12000;
 
 export class ApiManager {
   private socket: WebSocket;
   private derivApi: TDerivApi;
   private pingInterval: NodeJS.Timer;
+  private reconnectInterval: NodeJS.Timer;
+  private websocket_connected: (connection_value) => boolean;
+  private websocket_authorize: (connection_value) => boolean;
```
Comment (ali-hosseini-deriv) [inline src/configs/websocket/index.ts:28]:
```suggestion
  private is_websocket_authorized: (connection_value) => boolean;
```

File: src/configs/websocket/index.ts [hunk 2]
```diff
@@ -15,12 +15,17 @@ export type TDerivApi = {
...
+let attempts = 10;
+const RECONNECT_INTERVAL = attempts * 10000;
 const PING_INTERVAL = 12000;
 
 export class ApiManager {
   private socket: WebSocket;
   private derivApi: TDerivApi;
   private pingInterval: NodeJS.Timer;
+  private reconnectInterval: NodeJS.Timer;
+  private websocket_connected: (connection_value) => boolean;
```
Comment (ali-hosseini-deriv) [inline src/configs/websocket/index.ts:27]:
```suggestion
  private is_websocket_connected: (connection_value) => boolean;
```

File: src/configs/websocket/index.ts [hunk 3]
```diff
@@ -51,7 +56,9 @@ export class ApiManager {
     return this.derivApi.subscribe(request) as Observable<TSocketResponse<T>>;
   }
 
-  public authorize(token: string) {
+  public authorize(token: string, setIsConnected, setIsAuthorized) {
+    this.websocket_connected = setIsConnected;
```
Comment (ali-hosseini-deriv) [inline src/configs/websocket/index.ts:60]:
```suggestion
    this.is_websocket_connected = setIsConnected;
```

File: src/configs/websocket/index.ts [hunk 4]
```diff
@@ -51,7 +56,9 @@ export class ApiManager {
...
+    this.websocket_authorize = setIsAuthorized;
```
Comment (ali-hosseini-deriv) [inline src/configs/websocket/index.ts:61]:
```suggestion
    this.is_websocket_authorized = setIsAuthorized;
```

File: src/configs/websocket/index.ts [hunk 5]
```diff
@@ -62,14 +69,28 @@ export class ApiManager {
     if (this.pingInterval) {
       clearInterval(this.pingInterval);
     }
+    if (this.reconnectInterval) {
+      clearInterval(this.reconnectInterval);
+    }
     this.socket.addEventListener('open', () => {
+      this.websocket_connected && this.websocket_connected(true);
```
Comment (ali-hosseini-deriv) [inline src/configs/websocket/index.ts:76]:
```suggestion
      this.is_websocket_connected?.(true);
```

File: src/configs/websocket/index.ts [hunk 6]
```diff
@@ -62,14 +69,28 @@ export class ApiManager {
...
       this.pingInterval = setInterval(() => {
         this.socket.send(JSON.stringify({ ping: 1 }));
       }, PING_INTERVAL);
     });
 
     this.socket.addEventListener('close', () => {
+      this.websocket_connected && this.websocket_connected(false);
```
Comment (ali-hosseini-deriv) [inline src/configs/websocket/index.ts:83]:
```suggestion
      this.is_websocket_connected?.(false);
```

File: src/configs/websocket/index.ts [hunk 7]
```diff
@@ -62,14 +69,28 @@ export class ApiManager {
...
+      this.websocket_authorize && this.websocket_authorize(false);
```
Comment (ali-hosseini-deriv) [inline src/configs/websocket/index.ts:84]:
```suggestion
      this.is_websocket_authorized?.(false);
```

**Review by ali-hosseini-deriv — CHANGES_REQUESTED** [review]

File: src/configs/websocket/index.ts [hunk 8]
```diff
@@ -62,14 +69,28 @@ export class ApiManager {
...
       clearInterval(this.pingInterval);
+      this.socket = null;
+      if (attempts > 0) {
+        this.reconnectInterval = setTimeout(this.init.bind(this), RECONNECT_INTERVAL);
+        attempts -= 1;
+      } else {
+        window.alert('server down!!!');
```
Comment (matin-deriv) [inline src/configs/websocket/index.ts:91]: I don't suggest having an alert but implementing an in-house error handler component.
Alert is blocking the execution of the rest of the code and not a good user experience!

File: src/configs/websocket/index.ts [hunk 9]
```diff
@@ -15,12 +15,17 @@ export type TDerivApi = {
   authorize: (requestData: AuthorizeRequest) => Promise<AuthorizeResponse>;
 };
 
+let attempts = 10;
+const RECONNECT_INTERVAL = attempts * 10000;
```
Comment (matin-deriv) [inline src/configs/websocket/index.ts:19]: Are these numbers part of the task's requirements?

File: src/configs/websocket/index.ts (hunk 8 above)
Comment (utkarsha-deriv) [reply src/configs/websocket/index.ts:91]: Yes Matin, i have already dropped a message on the squad channel regarding as the requirement is not really clear on this. If we need to display message then we will need a design for it. I added this window alert as a temporary replacement to check the flow.

File: src/configs/websocket/index.ts (hunk 9 above)
Comment (utkarsha-deriv) [reply src/configs/websocket/index.ts:19]: It was not mentioned precisely, I saw in deriv-app we are reconnecting with a constant interval of 5000ms.

**Review by ali-hosseini-deriv — DISMISSED** [review]
LGTM

**Review by matin-deriv — APPROVED** [review]

**Review by amina-deriv — APPROVED** [review]

File: src/configs/websocket/index.ts [hunk 10]
```diff
@@ -88,7 +88,9 @@ export class ApiManager {
         this.reconnectInterval = setTimeout(this.init.bind(this), RECONNECT_INTERVAL);
         attempts -= 1;
       } else {
-        window.alert('server down!!!');
+        window.alert(
+          'Sorry, the server is currently down. Please refresh the page or try again later',
+        );
```
Comment (amina-deriv) [inline src/configs/websocket/index.ts:93]: Are we still shpwing alert message?

Comment (utkarsha-deriv) [reply src/configs/websocket/index.ts:93]: yes

**Review by matin-deriv — APPROVED** [review]

**Pull request merged**